lib/core/managementClient.ts

```diff
@@ -532,6 +556,224 @@ export class ManagementClient extends LinkEntity {
     }
   }
 
+
+  /**
+   * Get all the rules on the Subscription.
+   * @returns Promise<RuleDescription> A list of rules.
+   */
+  async getRules(): Promise<RuleDescription[]> {
+    try {
+      const request: AmqpMessage = {
+        body: {
+          top: types.wrap_int(max32BitNumber),
+          skip: types.wrap_int(0)
+        },
+        reply_to: this.replyTo,
+        application_properties: {
+          operation: Constants.operations.enumerateRules
+        }
+      };
+      request.application_properties![Constants.trackingId] = generate_uuid();
+
+      log.mgmt("[%s] Get rules request body: %O.", this._context.namespace.connectionId,
+        request.body);
+      log.mgmt("[%s] Acquiring lock to get the management req res link.",
+        this._context.namespace.connectionId);
+      await defaultLock.acquire(this.managementLock, () => { return this._init(); });
+
+      const response = await this._mgmtReqResLink!.sendRequest(request);
+      if (response.application_properties!.statusCode === 204 || !response.body || !Array.isArray(response.body.rules)) {
+        return [];
+      }
+
+      // Reference: https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-request-response#response-11
+      const result: { "rule-description": Typed }[] = response.body.rules || [];
+      const rules: RuleDescription[] = [];
+      result.forEach((x) => {
+        const ruleDescriptor = x['rule-description'];
+
+        // We use the first three elements of the `ruleDescriptor.value` to get filter, action, name
+        if (!ruleDescriptor
+          || !ruleDescriptor.descriptor
+          || ruleDescriptor.descriptor.value !== descriptorCodes.ruleDescriptionList
+          || !Array.isArray(ruleDescriptor.value)
+          || ruleDescriptor.value.length < 3) {
+          return;
+        }
+
+        const filtersRawData: Typed = ruleDescriptor.value[0];
+        const actionsRawData: Typed = ruleDescriptor.value[1];
+        const rule: RuleDescription = {
+          name: ruleDescriptor.value[2].value
+        };
+
+        switch (filtersRawData.descriptor.value) {
+          case descriptorCodes.trueFilterList:
+            rule.filter = {
+              expression: "1=1"
+            };
+            break;
+          case descriptorCodes.falseFilterList:
+            rule.filter = {
+              expression: "1=0"
+            };
+            break;
+          case descriptorCodes.sqlFilterList:
+            if (Array.isArray(filtersRawData.value) && filtersRawData.value.length === 2) {
+              rule.filter = {
+                expression: this._safelyGetTypedValue(filtersRawData.value[0])
+              };
+            }
+            break;
+          case descriptorCodes.correlationFilterList:
+            if (Array.isArray(filtersRawData.value) && filtersRawData.value.length === 9) {
```

Comment:
Is it guaranteed to have 9 properties in response (at all the times) even when all the properties are not defined? If not then we will lose the opportunity to deserialize it as a correlationFilterList.

If it does not have 9 properties then we should definitely log an error  (and may be throw an error?)